### PR TITLE
[ML] Update reindexing task progress before persisting job progress

### LINKED ISSUE #61852: RunDataFrameAnalyticsIT failures
There are a few tests that fail: 

org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithEnoughDocumentsToScroll
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithFewDocuments
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithMultipleSourceIndices
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithPreExistingDestIndex

with the same error. 

**Build scan**:
[7.10.0] https://gradle-enterprise.elastic.co/s/dp2u6q52kyxus
[7.9.1] https://gradle-enterprise.elastic.co/s/dlgajwjbzcmli
[7.10.0] https://gradle-enterprise.elastic.co/s/w57smywps3wlq
[7.9.1] https://gradle-enterprise.elastic.co/s/at3whtu5np6r6
https://gradle-enterprise.elastic.co/s/c5egkz647juis
 
and

[7.9.1] https://gradle-enterprise.elastic.co/s/jbaq7v3utks6q
[7.10.0] https://gradle-enterprise.elastic.co/s/bnbd27r4httos
[7.10.0] https://gradle-enterprise.elastic.co/s/si6k6omt5gjfe
[7.9.1] https://gradle-enterprise.elastic.co/s/blgn5kzzmp7pg

[7.10.0] https://gradle-enterprise.elastic.co/s/d5w6ctq3li7va
https://gradle-enterprise.elastic.co/s/3o4d53arbbfia

**Repro line**:

Example: 
```
REPRODUCE WITH: gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:integTest' --tests "org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testOutlierDetectionWithFewDocuments" \
  -Dtests.seed=E6859593F0B2A900 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-EC \
  -Dtests.timezone=Canada/Newfoundland \
  -Druntime.java=11
```

**Reproduces locally?**:
Didn't try ( windows)

**Applicable branches**:
master

**Failure history**:

FAiled twice today but hasnt been failing recently with this assertion error. 

**Failure excerpt**:

```

org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT > testOutlierDetectionWithFewDocuments FAILED
    java.lang.AssertionError: progress is complete: [{"phase":"reindexing","progress_percent":1}, {"phase":"loading_data","progress_percent":100}, {"phase":"computing_outliers","progress_percent":100}, {"phase":"writing_results","progress_percent":100}]
    Expected: is <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([E6859593F0B2A900:58D5761152F12CEA]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.assertProgressComplete(MlNativeDataFrameAnalyticsIntegTestCase.java:227)
        at org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testOutlierDetectionWithFewDocuments(RunDataFrameAnalyticsIT.java:143)
```

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This fixes a bug introduced by #61782. In that PR I thought I could
simplify the persistence of progress by using the progress straight
from the stats holder in the task instead of calling the get
stats action. However, I overlooked that it is then possible to
have stale progress for the reindexing task as that is only updated
when the get stats API is called.

In this commit this is fixed by updating reindexing task progress
before persisting the job progress. This seems to be much more
lightweight than calling the get stats request.

Closes #61852

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java [hunk 1]
```diff
@@ -331,14 +331,26 @@ void persistProgress(Client client, String jobId, Runnable runnable) {
...
+            },
+            e -> {
+                LOGGER.error(new ParameterizedMessage(
+                    "[{}] cannot persist progress as an error occurred while updating reindexing task progress", taskParams.getId()), e);
+                runnable.run();
+            }
+        );
+
+        // Step 1: Update reindexing progress as it could be stale
+        updateReindexTaskProgress(reindexProgressUpdateListener);
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java:353]: I was going to comment about "what else could be stale", but looking at how we handle the `_stats` call, we effectively do this.

So, we are just shortcutting the `_stats` call here by slightly "duplicating" the code here 

I am referencing:

https://github.com/elastic/elasticsearch/blob/c5cef17bd597eb42164e518c94308b5093025107/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java#L122

**Review by benwtrent — APPROVED** [review]
Its a bit weird that reindex is the only thing that could be stale. 

But I suppose it makes sense as it is the only "task" that we kick off where we don't have an internal monitor running.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java:353]: Yes indeed. Reindex is the only async task that we need to take care of updating its progress before reporting it. I thought about adding back the call to the get stats API. But it is quite a long way around. What do you think is better?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java:353]: I think doing this directly is fine.

We should definitely comment here AND on the _stats call. That way if ever we introduce another async service that we have to worry about refreshing, we capture it.

**Pull request merged**